### ISSUE TITLE
Addition: Some preliminary checks for while and if statements

### ISSUE THREAD

**Issue by Nokel81** [rust-lang-nursery/rust-clippy#2054]
This is a suggestion for addition for checks to `while` statements and `if` statements. This would mainly include checks for things like tautologies (similar to the lint for `while` to `loop`) and overlapping check statements in `if`/`else if` statements.

**Comment by Nokel81** [comment]
How does this sound, if it sounds good then I would be fine implementing it and making a PR

**Comment by oli-obk** [comment]
We already have such a lint. It's called `logic_bug`. A related lint is `nonminimal_bool`. Please give some examples that aren't covered by those lints so we can discuss how to improve them.

**Comment by oli-obk** [comment]
Open issues that I can think of:

# reusing the same condition in later conditions

```rust
    if x == 6 {
        println!("bar")
    } else if x == 6 && y == 7 {
        println!("foo");
    }
```

# comparing the same variable in one expression

```rust
x == 6 && x == 7 // never true
```

```rust
x == 6 && x != 7 // x == 6 implies x != 7
```

**Comment by Nokel81** [comment]
With the comparing of the same variable in one expression there are a few things
1. If the comparison has `<` or `>` in it then it should be all going the same way
```rust
x >= 3 && x < 10 // inconsistent ordering 
```
should be 
```rust
3 <= x && x < 10
```
1. Should also include range checking for types.
```rust
let x:u32 = 10;
...
x > -1 // always true
x < 0 // always false
```
1. Checking if the bounds of a range are possible 
```rust
x > 10 && x < 6
```

**Comment by clarcharr** [comment]
1. This should have a check that the right `PartialEq` implementations are there.
2. Wouldn't `x > -1` compile-fail for `u32`
3. Makes sense.

**Comment by Nokel81** [comment]
Can clippy check if the implementation is correct?

**Comment by oli-obk** [comment]
If you implement bogus PartialEq, that's either impossible to lint or at least really hard. open a different issue for that. In this lint we need to assume the impls of the different comparison traits agree
